File: docs/code-quality/codesnippet/CPP/ca2227-collection-properties-should-be-read-only_1.cpp
#include "stdafx.h"
using namespace System;
using namespace System::Collections;

namespace UsageLibrary
{
	public ref class WritableCollection
	{
	public:
		property ArrayList^ SomeStrings
		{
			ArrayList^ get() { return someStrings; }

			// This set accessor violates rule CA2227.
			// To fix the code, remove this set accessor.
			void set(ArrayList^ value) { someStrings = value; }
		}

		WritableCollection()
		{
			someStrings = gcnew ArrayList(gcnew array<String^> {"one", "two", "three"});
		}

	private:
		ArrayList ^ someStrings;
	};
}

using namespace UsageLibrary;

void main()
{
	ArrayList^ newCollection = gcnew ArrayList(gcnew array<String^> {"a", "new", "collection"});

	WritableCollection^ collection = gcnew WritableCollection();

	// This line of code demonstrates how the entire collection
	// can be replaced by a property that's not read only.
	collection->SomeStrings = newCollection;

	// If the intent is to replace an entire collection,
	// implement and/or use the Clear() and AddRange() methods instead.
	collection->SomeStrings->Clear();
	collection->SomeStrings->AddRange(newCollection);
}

File: docs/code-quality/codesnippet/CSharp/ca2227-collection-properties-should-be-read-only_1.cs
using System;
using System.Collections;

namespace UsageLibrary
{
   public class WritableCollection
   {
      ArrayList strings;

      public ArrayList SomeStrings
      {
         get { return strings; }

         // This property violates rule CA2227.
         set { strings = value; }
      }

      public WritableCollection()
      {
         strings = new ArrayList(
            new string[] {"IEnumerable", "ICollection", "IList"} );
      }
   }

   class ReplaceWritableCollection
   {
      static void Main()
      {
         ArrayList newCollection = new ArrayList(
            new string[] {"a", "new", "collection"} );

         WritableCollection collection = new WritableCollection();
         // This line of code demonstrates how the entire collection
         // can be replaced by a property that's not read-only.
         collection.SomeStrings = newCollection;

         // If the intent is to replace an entire collection,
         // implement Clear() and AddRange() methods instead.
         collection.SomeStrings.Clear();
         collection.SomeStrings.AddRange(newCollection);
      }
   }
}